iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java

```diff
@@ -1817,6 +1821,8 @@ public boolean canUseTruncate(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
     Table table = IcebergTableUtil.getTable(conf, hmsTable.getTTable());
     if (MapUtils.isEmpty(partitionSpec) || !hasUndergonePartitionEvolution(table)) {
       return true;
+    } else if (hmsTable.getSnapshotRef() != null) {
```

Comment:
why can't we use truncate when branch is provided?

Reply:
this is branch is involved & the table has undergone partition evolution, if it didn't went through partition evolution the above if will return true for branch as well

Reply:
oh, but why do we need to check the branch name if table has undergone partition evolution, shouldn't we fallback right away?

Reply:
below there is some more optimisation for normal table, I don't catch it well but it is checking if the table has undergone partition evolution, but there is no datafile outside the new partition or so, so in that case even in case of partition evolution you can go with truncate but that would throw you an NPE, in case of branches if the partition column is removed in the current table spec.